database.py

```diff
@@ -402,7 +407,11 @@ def refresh_user_best_runs(
         Flushes the best_runs table for a given user and year-day-part, used when inserting new entries or when marking submissions invalid
         """
         self._cursor.execute(
-            "REPLACE INTO best_runs "
+            "DELETE FROM best_runs WHERE (year = ? AND day_part = ? AND user = ?)",
+            (year, pack_day_part(day, part), user_id),
+        )
+        self._cursor.execute(
+            "INSERT INTO best_runs "
             + "SELECT user, year, day_part, MIN(average_time) AS best_time, submission_id AS run_id FROM submissions "
             + "WHERE (year = ? AND day_part = ? AND valid = 1 AND user = ?)",
             (year, pack_day_part(day, part), user_id),
```

Comment:
REPLACE INTO works row-by-row. It's an alias for INSERT OR REPLACE. That doesn't work for when a submission gets invalidated; the DB attempts to match rows by the Submission ID. What we're actually going for is a replacement of one set of rows with another set of rows, not replacing individual rows.

Reply:
oops, i forgot the unique constraint we had got updated to include best_time aswell, in a previous implementation replace would work here